src/Neo.VM/JumpTable/JumpTable.Compound.cs

```diff
@@ -280,10 +280,15 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
             // For arrays, check if the index is within bounds and push the result onto the stack.
             case VMArray array:
                 {
-                    // TODO: Overflow and underflow checking needs to be done.
-                    var index = (int)key.GetInteger();
-                    if (index < 0)
-                        throw new InvalidOperationException($"The negative index {index} is invalid for OpCode.{instruction.OpCode}.");
+                    var indexBI = key.GetInteger();
+                    if (indexBI.Sign < 0)
+                        throw new InvalidOperationException($"The negative index {indexBI} is invalid for OpCode.{instruction.OpCode}.");
+                    if (indexBI > int.MaxValue)
+                    {
+                        engine.Push(false);
+                        break;
+                    }
+                    var index = (int)indexBI;
                     engine.Push(index < array.Count);
```

Comment:
```suggestion
                    engine.Push(indexBI < array.Count);
```
It's possible to compare directly without previous cast